File: src/main/java/dev/zprestige/fire/module/combat/autocrystal/PacketSendListener.java
package dev.zprestige.fire.module.combat.autocrystal;

import dev.zprestige.fire.event.bus.EventListener;
import dev.zprestige.fire.event.impl.PacketEvent;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityEnderCrystal;
import net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock;
import net.minecraft.util.math.BlockPos;

import java.util.ConcurrentModificationException;

public class PacketSendListener extends EventListener<PacketEvent.PacketSendEvent, AutoCrystal> {

    public PacketSendListener(final AutoCrystal autoCrystal) {
        super(PacketEvent.PacketSendEvent.class, autoCrystal);
    }

    @Override
    public void invoke(final Object object) {
        final PacketEvent.PacketSendEvent event = (PacketEvent.PacketSendEvent) object;
        if (event.getPacket() instanceof CPacketPlayerTryUseItemOnBlock) {
            if (module.predict.GetCombo().equals("Ultra") || module.predict.GetCombo().equals("UltraChain")) {
                final BlockPos eventPos = ((CPacketPlayerTryUseItemOnBlock) event.getPacket()).getPos();
                EntityEnderCrystal entityEnderCrystal = null;
                try {
                    for (Entity entity : mc.world.loadedEntityList) {
                        if (entity instanceof EntityEnderCrystal && entity.getDistanceSq(eventPos.up()) < 1.0f) {
                            entityEnderCrystal = (EntityEnderCrystal) entity;
                        }
                    }
                } catch (ConcurrentModificationException ignored) {
                }
                if (entityEnderCrystal != null) {
                    module.explodeCrystal(entityEnderCrystal, null);
                    if (module.predict.GetCombo().equals("UltraChain")) {
                        module.placeCrystal(eventPos, null);
                    }
                }
            }
        }
    }
}

File: src/main/java/dev/zprestige/fire/module/misc/removeexpdesync/PacketSendListener.kt
package dev.zprestige.fire.module.misc.removeexpdesync

import dev.zprestige.fire.event.bus.EventListener
import dev.zprestige.fire.event.impl.PacketEvent.PacketSendEvent
import net.minecraft.network.play.client.CPacketPlayerTryUseItem
import net.minecraft.network.play.client.CPacketPlayerTryUseItemOnBlock

class PacketSendListener(removeEXPDesync: RemoveEXPDesync) : EventListener<PacketSendEvent, RemoveEXPDesync>(PacketSendEvent::class.java, removeEXPDesync) {
    
    override fun invoke(e: Any) {
        val event = e as PacketSendEvent
        if (event.packet is CPacketPlayerTryUseItem) {
            val packet = event.packet as CPacketPlayerTryUseItem
            if (module.isEXP(packet.hand)) {
                module.started = true
            }
        }
        if (event.packet is CPacketPlayerTryUseItemOnBlock) {
            val packet = event.packet as CPacketPlayerTryUseItemOnBlock
            if (module.isEXP(packet.getHand())) {
                module.started = true
            }
        }
    }
}